### PR TITLE
8313779: RISC-V: use andn / orn in the MD5 instrinsic

### PR DESCRIPTION
**Pull request by gctony**
Small improvement of the MD5 intrinsic when the Zbb extension is available. Performance comparison (with thanks again to @robehn for this!):

before:

```
MessageDigests.digest                   md5        64     DEFAULT  avgt    6    1835.246 ±  252.071  ns/op
MessageDigests.digest                   md5     16384     DEFAULT  avgt    6  145386.522 ±  444.446  ns/op
MessageDigests.getAndDigest             md5        64     DEFAULT  avgt    6    2555.515 ±  639.491  ns/op
MessageDigests.getAndDigest             md5     16384     DEFAULT  avgt    6  149045.631 ± 6658.545  ns/op
```

after:

```
MessageDigests.digest                   md5        64     DEFAULT  avgt    6    1779.637 ±  207.869  ns/op
MessageDigests.digest                   md5     16384     DEFAULT  avgt    6  137147.179 ±  706.396  ns/op
MessageDigests.getAndDigest             md5        64     DEFAULT  avgt    6    2645.354 ± 1245.318  ns/op
MessageDigests.getAndDigest             md5     16384     DEFAULT  avgt    6  141306.966 ± 7000.576  ns/op
```

(only line 3 is not an improvement, but it has higher variation)

It seems to save around 5% of executed instructions.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313779](https://bugs.openjdk.org/browse/JDK-8313779): RISC-V: use andn / orn in the MD5 instrinsic (**Enhancement** - P4)

### Reviewers
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer) ⚠️ Review applies to [d7d1cfee](https://git.openjdk.org/jdk/pull/15156/files/d7d1cfeec3adabc91acaec16671d597294a691f1)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) ⚠️ Review applies to [d7d1cfee](https://git.openjdk.org/jdk/pull/15156/files/d7d1cfeec3adabc91acaec16671d597294a691f1)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15156/head:pull/15156` \
`$ git checkout pull/15156`

Update a local copy of the PR: \
`$ git checkout pull/15156` \
`$ git pull https://git.openjdk.org/jdk.git pull/15156/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15156`

View PR using the GUI difftool: \
`$ git pr show -t 15156`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15156.diff">https://git.openjdk.org/jdk/pull/15156.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15156#issuecomment-1665595293)

### CODE REVIEW

**Comment by gctony** [comment]
Was not sure how to differentiate between the `andn` and `orn` in the assembler and the macro assembler. I added an `r` suffix. If there's a better way, let me know!

**Comment by gctony** [comment]
correctness testing:

```
test/jdk/java/security/MessageDigest
test/hotspot/jtreg/compiler/intrinsics/sha/sanity
```

with `-UseZbb` and `+UseZbb`

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 1]
```diff
@@ -1654,6 +1654,28 @@ void MacroAssembler::xorrw(Register Rd, Register Rs1, Register Rs2) {
   sign_extend(Rd, Rd, 32);
 }
 
+// Rd = Rs1 & (~Rd2)
+void MacroAssembler::andnr(Register Rd, Register Rs1, Register Rs2) {
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:1658]: Keep the name `andn` here, and at https://github.com/openjdk/jdk/pull/15156/files#diff-7a5c3ed05b6f3f06ed1c59f5fc2a14ec566a6a5bd1d09606115767daa99115bdR1660 use `Assembler::andn`.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 2]
```diff
@@ -1654,6 +1654,28 @@ void MacroAssembler::xorrw(Register Rd, Register Rs1, Register Rs2) {
...
+  }
+
+  notr(Rd, Rs2);
+  andr(Rd, Rs1, Rd);
+}
+
+// Rd = Rs1 | (~Rd2)
+void MacroAssembler::ornr(Register Rd, Register Rs1, Register Rs2) {
+  if (UseZbb) {
+    orn(Rd, Rs1, Rs2);
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:1671]: Same as above.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 3]
```diff
@@ -4032,9 +4030,8 @@ class StubGenerator: public StubCodeGenerator {
               int k, int s, int t,
               Register rtmp1, Register rtmp2, Register rmask32) {
     // rtmp1 = c ^ (b | (~d))
-    __ notr(rtmp2, d);
-    __ orr(rtmp1, b, rtmp2);
-    __ xorr(rtmp1, c, rtmp1);
+    __ ornr(rtmp2, b, d);
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:4033]: Nit: You could use `rtmp1` here, the dependency for the next instruction will block the CPU from executing OOO anyway.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 1 above)
Comment (gctony) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Thanks. See updated patch.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 2 above)
Comment (gctony) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Ditto.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp (hunk 3 above)
Comment (gctony) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp]: Done.

**Review by luhenry — APPROVED** [review]

**Review by RealFYang — APPROVED** [review]
Looks good.
BTW: I noticed that two `li` are used in your previous PR to implement MD5 intrinsic:
```
./stubGenerator_riscv.cpp:    __ li(rtmp2, t);
./stubGenerator_riscv.cpp:    __ li(rmask32, MASK_32);
```
I think It should be more consistent if you use 'mv' instead as we did in other places when moving an immediate into register. 
```
template<typename T, ENABLE_IF(std::is_integral<T>::value)>
inline void mv(Register Rd, T o)                    { li(Rd, (int64_t)o); }
```

**Comment by gctony** [comment]
> I think It should be more consistent if you use 'mv' instead as we did in other places when moving an immediate into register.

Thanks for the suggestion! I thought `mv` was for register to register copy. But I'll change `li` to `mv` in those two places. I'll integrate the change as soon as I retest.

**Comment by gctony** [comment]
Tests pass (with both `-UseZbb` and `+UseZbb`) with the `li` -> `mv` change. I also confirmed that the generated code is the same.

**Pull request closed without merging**